### PR TITLE
Migrate java scopes

### LINKED ISSUE #1926: Java: Support name and value in for each statements
First requires name and value to migrate from legacy scope

reference https://github.com/cursorless-dev/cursorless/pull/1919

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
#Fixes #1926

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [-] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [-] I have not broken the cheatsheet

### CODE REVIEW

File: queries/java.scm [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+;;!  ------------------------------
+(for_statement
+    condition: (_) @condition
+) @_.domain
+
+;;!! switch (value) {}
+;;!          ^^^^^
+;;!  -----------------
+(switch_expression
+    condition: (_) @condition @switchStatementSubject
```
Comment (AndreasArvidsson) [inline queries/java.scm:124]: We seem to have a conflict here. I just preserve the existing behavior, but this should probably change.

File: queries/java.scm [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+) @namedFunction @functionName.domain
+(constructor_declaration
+    (identifier) @functionName
+) @namedFunction @functionName.domain
+
+(lambda_expression) @anonymousFunction
+
+(
+    (if_statement) @ifStatement
+    (#not-parent-type? @ifStatement "if_statement")
```
Comment (pokey) [inline queries/java.scm:54]: This doesn't match what was there before. I'm assuming this is an improvement, but worth adding a test to capture the new behaviour if so

File: queries/java.scm [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+;;!! if () {}
+;;!  ^^^^^^^^
+(
+    (if_statement
+        "if" @branch.start @condition.domain.start
+        condition: (_) @condition
+        consequence: (block) @branch.end @condition.domain.end
+    ) @dummy
+    (#not-parent-type? @dummy "if_statement")
+    (#child-range! @condition 0 -1 true true)
```
Comment (pokey) [inline queries/java.scm:92]: This includes any whitespace surrounding the condition inside the parens. Probably fine. Def makes sense for removal range. Prob good enough for content range

File: queries/java.scm (hunk 1 above)
Comment (pokey) [reply queries/java.scm:124]: Does a test break if you drop `@condition` here? If not, I'd be tempted to drop it

**Review by pokey — COMMENTED** [review]
Looks good! Left a few minor comments

File: queries/java.scm (hunk 2 above)
Comment (AndreasArvidsson) [reply queries/java.scm:54]: done

File: queries/java.scm (hunk 1 above)
Comment (AndreasArvidsson) [reply queries/java.scm:124]: Seem to work without it. But currently there is no support for condition in switch matches. This doesn't seem to work in typescript either. Is the point that condition should be the values in the match statement?

File: queries/java.scm (hunk 3 above)
Comment (AndreasArvidsson) [reply queries/java.scm:92]: I think so

File: queries/java.scm [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    (#child-range! @switchStatementSubject 0 -1 true true)
+) @_.domain
+
+;;!! true ? 1 : 2
+(ternary_expression
+    condition: (_) @condition
+    consequence: (_) @branch
+) @condition.domain
+(ternary_expression
+    alternative: (_) @branch
```
Comment (AndreasArvidsson) [inline queries/java.scm:160]: Do we have a way to define a scope multiple places in the same pattern?

Comment (pokey) [reply queries/java.scm:191]: unfortunately not. The problem is that then things like `.leading`, `.start`, etc become ambiguous. I guess we could allow it if we don't see any of those things in the pattern? 🤔 cc/ @josharian

File: queries/java.scm [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+;;!! new test()
+;;!  ^^^^^^^^
+;;!  ----------
+(object_creation_expression
+    (argument_list) @functionCallee.end.startOf
+) @functionCallee.start.startOf @_.domain
+
+;;!! new test().bar();
+;;!  ^^^^^^^^^^^^^^
+;;!  ----------------
```
Comment (pokey) [inline queries/java.scm:264]: does this actually exclude the `;`? If so, we should prob try to include that

File: queries/java.scm [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+(method_invocation
+    (argument_list) @functionCallee.end.startOf
+) @functionCallee.start.startOf @_.domain
+
+;;!! super();
+;;!  ^^^^^
+;;!  -------
```
Comment (pokey) [inline queries/java.scm:271]: as above

**Review by pokey — COMMENTED** [review]
Ok I didn't look super closely, as it's been awhile since my java days, we have lots of tests, and you're using it day-to-day so will notice oddnesses.  I left a few comments. In addition, the following looked weird to me for "value"

image

domain should include `;` above; might be same issue I point out in inline comment

image

these should be values

File: queries/java.scm (hunk 5 above)
Comment (AndreasArvidsson) [reply queries/java.scm]: fixed

File: queries/java.scm (hunk 6 above)
Comment (AndreasArvidsson) [reply queries/java.scm]: fixed

**Comment by AndreasArvidsson** [comment]
@pokey Updated.

File: queries/java.scm (hunk 4 above)
Comment (josharian) [reply queries/java.scm:192]: I have now forgotten all of the details but last time I tried playing games like that I ended up in a tangled mess of code. I will generally choose clear over clever/concise, personally.

Comment (AndreasArvidsson) [reply queries/java.scm:192]: Couldn't we number the scopes? 
```
(ternary_expression
    consequence: (_) @branch1
    alternative: (_) @branch2
)
```

If we can't use actual numbers we could use letters or roman numerals? :D

Comment (pokey) [reply queries/java.scm:192]: That's an interesting idea. Or `@branch.1`

Comment (AndreasArvidsson) [reply queries/java.scm:192]: Yeah probably even better

File: queries/java.scm [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+;;!          ^
+;;!       xxxx
+;;!  ----------
+(_
+    (assignment_expression
+        left: (_) @value.removal.start.endOf
+        right: (_) @value @value.removal.end
+    ) @_.domain.start
+    ";"? @_.domain.end
+)
```
Comment (pokey) [inline queries/java.scm:322]: I'd argue for using leading / trailing instead of removal for all of these

**Review by pokey — COMMENTED** [review]
Better; left one more minor comment and I think we're good to go

File: queries/java.scm (hunk 7 above)
Comment (AndreasArvidsson) [reply queries/java.scm:322]: If you prefer I could do that. Personally I don't think I have every use leading and trailing for anything else than white spaces, but I'm not against it.

Comment (AndreasArvidsson) [reply queries/java.scm]: fixed

Comment (pokey) [reply queries/java.scm]: Ha yeah was kind of thinking the same thing in the back of my head 😅. Better to be consistent tho I guess 🤷‍♂️

**Review by pokey — APPROVED** [review]
👍👍

**Pull request merged**